bayeux.go

```diff
@@ -298,6 +332,7 @@ func GetSalesforceCredentials(ap AuthenticationParameters) (creds *Credentials,
 
 func (b *Bayeux) Channel(ctx context.Context, out chan MaybeMsg, r string, creds Credentials, channel string) chan MaybeMsg {
 	b.creds = creds
+	b.status = status
```

Comment:
Could you please explain what are you trying to achieve? In L132 the status is global (which is not good in general), but here you're setting it as part of the `Bayeux`?

Reply:
I have added  one parameter `connectCount` in Status struct. I need to ensure if connections are closed as soon as context is canceled from filebeat side(for test case to ensure if channel is closing or not). To do that I need some mechanism in bayeux. but i don't have any way to access status object so for reference I used bayeux object.

Reply:
Are you sure that this code doesn't need any concurrency safety mechanism? For example in [MQTT input](https://github.com/elastic/beats/blob/main/filebeat/input/mqtt/input.go#L51), I used a WaitGroup.

Reply:
What do you think of this for concurrency safety mechanism for in flight messages. as we can only close channel from bayeux side. we will wait till all the inbound messages are sent via channel.
```var waitMsgs sync.WaitGroup
	wg.Add(1)
	go func() {
		defer func() {
			waitMsgs.Wait()
			close(out)
			status.disconnect()
			wg.Done()
		}()
		for {
			for _, e := range x {
				waitMsgs.Add(1)
				go func() {
					defer waitMsgs.Done()
					// close channel as soon as context is canceled to ensure it won't hold channel open for forever
					out <- MaybeMsg{Msg: e}
				}()
			}
		}
	}()
	return out
```

Reply:
It's hard for me to tell, as the entire library code is tangled a bit. I can't confirm that this change doesn't introduce any other issues (deadlocks). What I don't like about the suggested approach is for-in-for loop.

WDYT about refactoring the library code or writing it from scratch? We may try to get rid of all globals and focus on concurrency safety.

Reply:
Oh, sorry about that, I removed other things there to make it simple to understand but actual snip should look like this:

```
func (b *Bayeux) connect(ctx context.Context, out chan MaybeMsg) chan MaybeMsg {
	var waitMsgs sync.WaitGroup
	wg.Add(1)
	go func() {
		defer func() {
			waitMsgs.Wait()
			close(out)
			status.disconnect()
			wg.Done()
		}()
		for {
			select {
			case <-ctx.Done():
				return
			default:
				postBody := fmt.Sprintf(`{"channel": "/meta/connect", "connectionType": "long-polling", "clientId": "%s"} `, b.id.clientID)
				resp, err := b.call(ctx, postBody, b.creds.bayeuxUrl())
				if err != nil {
					if errors.Is(err, context.Canceled) {
						return
					}
					out <- MaybeMsg{Err: fmt.Errorf("cannot connect to bayeux: %s, trying again", err)}
				} else {
					if os.Getenv("DEBUG") != "" {
						var b []byte
						if resp.Body != nil {
							b, _ = ioutil.ReadAll(resp.Body)
						}
						// Restore the io.ReadCloser to its original state
						resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
						// Use the content
						s := string(b)
						logger.Printf("Response Body: %s", s)
					}
					var x []TriggerEvent
					decoder := json.NewDecoder(resp.Body)
					if err := decoder.Decode(&x); err != nil && err == io.EOF {
						out <- MaybeMsg{Err: err}
						return
					}
					for _, e := range x {
						waitMsgs.Add(1)
						go func() {
							defer waitMsgs.Done()
							// close channel as soon as context is canceled to ensure it won't hold channel open for forever
							select {
							case <-ctx.Done():
								return
							default:
								out <- MaybeMsg{Msg: e}
							}
						}()
					}
				}
			}
		}
	}()
	return out
}
```

Reply:
In theory, it looks good to me, let's get it into the PR code.